src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+                mir: &mut Mir<'tcx>) {
+    let gen_ty = mir.local_decls.raw[1].ty;
+
+    let region = ty::ReFree(ty::FreeRegion {
+        scope: def_id,
+        bound_region: ty::BoundRegion::BrEnv,
+    });
+
+    let region = tcx.mk_region(region);
+
+    let ref_gen_ty = tcx.mk_ref(region, ty::TypeAndMut {
+        ty: gen_ty,
+        mutbl: hir::MutMutable
+    });
+
+    // Replace the by value generator argument
+    mir.local_decls.raw[1].ty = ref_gen_ty;
+
+    // Add a deref to accesses of the generator state
+    DerefArgVisitor.visit_mir(mir);
+}
+
+fn replace_result_variable<'tcx>(ret_ty: Ty<'tcx>,
+                            mir: &mut Mir<'tcx>) -> Local {
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let new_ret = LocalDecl {
+        mutability: Mutability::Mut,
+        ty: ret_ty,
+        name: None,
+        source_info,
+        internal: false,
+        is_user_variable: false,
+    };
+    let new_ret_local = Local::new(mir.local_decls.len());
+    mir.local_decls.push(new_ret);
+    mir.local_decls.swap(0, new_ret_local.index());
+
+    RenameLocalVisitor {
+        from: RETURN_POINTER,
+        to: new_ret_local,
+    }.visit_mir(mir);
+
+    new_ret_local
+}
+
+fn locals_live_across_suspend_points<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                               mir: &Mir<'tcx>,
+                                               source: MirSource) -> liveness::LocalSet {
+    use rustc_data_structures::indexed_set::IdxSetBuf;
+    let mut set = liveness::LocalSet::new_empty(mir.local_decls.len());
+    let result = liveness::liveness_of_locals(mir);
+    liveness::dump_mir(tcx, "generator_liveness", source, mir, &result);
+
+    for (block, data) in mir.basic_blocks().iter_enumerated() {
+        if let TerminatorKind::Yield { .. } = data.terminator().kind {
+            set.union(&result.outs[block]);
+        }
+    }
+
+    // The generator argument is ignored
+    set.remove(&Local::new(1));
+
+    set
+}
+
+fn compute_layout<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                            def_id: DefId,
+                            source: MirSource,
+                            interior: GeneratorInterior<'tcx>,
+                            mir: &mut Mir<'tcx>)
+    -> (HashMap<Local, (Ty<'tcx>, usize)>, GeneratorLayout<'tcx>)
+{
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let mut live_locals = locals_live_across_suspend_points(tcx, mir, source);
+
+    let allowed = tcx.erase_regions(&interior.as_slice());
+
+    for (local, decl) in mir.local_decls.iter_enumerated() {
+        if !live_locals.contains(&local) || decl.internal {
+            continue;
+        }
+        if !allowed.contains(&decl.ty) {
+            tcx.sess.span_warn(mir.span,
+                &format!("generator contains type {} in MIR, but typeck only knows about {}",
```

Comment:
Then it would be nice if you would ad  `Broken MIR` or something to the warning, `transform::type_check` style.